### PR TITLE
[Minor] Add ExchangeRateService

### PR DESCRIPTION
**Pull request by 4hebek**
Add a service to get the current daily exchange rates from https://api.cnb.cz/cnbapi/exrates/daily

### CODE REVIEW

File: BackendTaskPilot/Program.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:10]: Since the CBN endpoint is a GET and is designed to retrieve data, we should use .MapGet instead. We should also add a meaningul resource name to it, perhaps `exchangerates` ( a small tip, resource names should be nouns) and pass the target currencies as query parameters. If you were using a separate Controller and inheriting the ControllerBase you'd use [FromQuery] for this. However for minimal APIs it seems to be different, have a look at the official documentation https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-7.0 (or change the url suffix to 8.0 if you decide to update the .net version of your project)

File: BackendTaskPilot/Program.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:16]: It might be a good idea to be more specific and rename Rates to ExchangeRates, this is just in case we ever expose this part in the .Contracts as a nuget package and consume it by other apps. This is so that we don't confuse exchange Rates with hotel pricing Rates in the future for example. Just a note, if we were to ever expose this as a package, the Output Type of this project should change from Console App to a Class Library.

File: BackendTaskPilot/Program.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+public struct Rate
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:22]: We use decimal instead of float for financial operations, it gives us more precision, it provides more decimal places

File: BackendTaskPilot/Program.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:27]: The url could be added to the appsettings file. I like how you made the variable name generic and not CNB specific as perhaps in the future we might change the api we use.

File: BackendTaskPilot/Program.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:33]: Since you're using the GetStringAsync method, you should make the whole GetExchangeRates method async and use the await keyword before the call. This is to make it asynchronous and stop it from blocking the thread. Here's quite a simple blog post that explains asynchronous programming https://auth0.com/blog/introduction-to-async-programming-in-csharp/

File: BackendTaskPilot/Program.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        var response = JsonSerializer.Deserialize<CnbRatesResponse>(data);
+
+        return response.Rates.Select(rate => rate).Where(rate => currencies.Contains(rate.CurrencyCode));
+    }
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:37]: Think about some logging and error handling. Add correlation tokens to the logs as this will help debugging especfially when we have other microservices calling this. Add some nulls checks and appropriate logs as you don't want to perform the rest of the work if there are no rates returned as it could lead to NullRefrenceExceptions downstream. Also when we request rates for multiple currencies we can add some logs of Warning type for the currencies that haven't been returned. As this would be quite confusing for whatever/whoever is callling this endpoint and expecting 5 different rates but only getting 3, logs would make this clearer. I have noticed that rates for some required currencies aren't being returned, perhaps will need to think of an alternative endpoint

**Comment by 4hebek** [comment]
We should definitely add some tests to this, you can use XUnit to do that. A few happy path test cases with the required currencies i.e. USD, EUR, JPY, SAR, KES individually as well as testing getting them all in one call will be the minimum. You can then start thinking about the edge cases and unhappy paths. These will act as foundations and help you programm your solution based on that/to fulfill those and cover all the exceptions mentioned above.

**Comment by 4hebek** [comment]
I'd think about the restructuring this work especially as it's most likely to grow since this is just the first iteration. We usually try to have a separate Services folder which would be suitable for your ExchangeRateService, we normally have a Models folder which is a good place for your CnbRatesResponse and Rate. Have a look at how it's done in www.github.com/Mews/ImaginaryProject. Restructing this will also help you to take some of the SOLID principles in the account.

File: BackendTaskPilot/Program.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:31]: Instantiating the HttpClient inside this method each time can be very resource consuming. It's better to extract it for various reasons. One being we want to stick to SOLID principles, for example the Single Responsibility Principle where the method should only have one reason to change. Having the client setup in a separate class also means we can perform more operations on it, for example setting the BaseAddress or the accepted DefaultRequestHeaders

File: BackendTaskPilot/Program.cs (hunk 2 above)
Comment (4hebek) [reply BackendTaskPilot/Program.cs:16]: Also using a struct here is quite confusing, structs are value types and the Rates property is a List<Rate> which is a reference type. Also structs don't support inheritance, so if we ever want to extend these by inheritance, we will not be able to do this. We use classes much more frequently https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

File: BackendTaskPilot/Program.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:7]: We should create a separate class for the ExchageRateService and create an interface for it and use Dependency Injection to instantiate it in the following way:
`builder.Services.AddSingleton<IExchangeRateService, ExchangeRateService>();`

This means the code that deals with getting and filtering the exchange rate is encapsulated and also makes the code more organised. Additionally this allows us to mock the service easier with the Moq library when we write our unit or integration tests.

File: BackendTaskPilot/Program.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:5]: Only these methods are good candidates for the use of a Program (or Startup class). The rest should be extracted in their separate folders and classes.

File: BackendTaskPilot/Program.cs (hunk 5 above)
Comment (4hebek) [reply BackendTaskPilot/Program.cs:33]: However it's also a good idea to use GetAsync instead and then Deserialize the .Content of the response. This means you can get the HttpResponseMessage object which will give you access to properties such as IsSuccessStatusCode or StatusCode. These can then help you handle some negative scenarios e.g. when no rates are found and you get a 404.

File: BackendTaskPilot/Program.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:29]: According to the requirements, we want a service that accepts both, a source (CZK as a default for now) and target currencies. I can see you're only dealing with target currencies here. Might be worth adding an optional parameter that defaults to CZK after the target currencies one. Although I've noticed the CNB endpoint only supports CZK as the default currency, which again might have to be addressed in the future, I'll speak to the Staff Engineer about a possible API alternative

File: BackendTaskPilot/Program.cs (hunk 3 above)
Comment (4hebek) [reply BackendTaskPilot/Program.cs:22]: It's best to create a Models folder for CnbRatesResponse and Rate and group them together

Comment (4hebek) [reply BackendTaskPilot/Program.cs:22]: Also we tend to have the [JsonPropertyName()] attribute on a separate line, on top of the property but that's just a style preference

File: BackendTaskPilot/Program.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        HttpClient client = new HttpClient();
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
+        var response = JsonSerializer.Deserialize<CnbRatesResponse>(data);
+
+        return response.Rates.Select(rate => rate).Where(rate => currencies.Contains(rate.CurrencyCode));
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:36]:
```suggestion
        return response.Rates.Where(rate => currencies.Contains(rate.CurrencyCode));
```

File: BackendTaskPilot/Program.cs (hunk 2 above)
Comment (4hebek) [reply BackendTaskPilot/Program.cs:16]: We can actually use an IEnumerable here instead, we only read and iterate over it so using a List for performance reasons shouldn't be the deciding factor.

File: BackendTaskPilot/Program.cs (hunk 10 above)
Comment (4hebek) [reply BackendTaskPilot/Program.cs:29]: Can also use an IEnumerable instead of a List here

**Comment by 4hebek** [comment]
A few things to address but a great job overall! Good use of minimal APIs, nice neat and readable solution, well done :) Let me know if you want to discuss any of the points in more detail, it's always easier to jump on a huddle on Slack.

File: BackendTaskPilot/BackendTaskPilot.csproj [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+    <PropertyGroup>
+        <TargetFramework>net7.0</TargetFramework>
```
Comment (4hebek) [inline BackendTaskPilot/BackendTaskPilot.csproj:4]: net8.0 is currently the latest active version so you could update the project to use that instead. However we have to be mindful of any dependencies and any projects that consume this one to ensure compatibility. Also it's possible that the build pipeline will need to be updated to accommodate the .net8.0 update. Stick any questions you have about CI/CD in the #imaginarydevops channel.

**Comment by 4hebek** [comment]
Please don't forget to update the readme, followed by all the documentation (e.g. on Confluence) as well as the Postman collection once this has been merged

File: BackendTaskPilot/Program.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
```
Comment (4hebek) [inline BackendTaskPilot/Program.cs:13]: Might be worth having a think about Swagger support, which is what we use to house our API documentation

**Comment by 4hebek** [comment]
In terms of folder structure, I'd think about restructuring this work especially as it's likely to grow since this is just the first iteration. We usually try to have a separate Services folder which would be suitable for some of the code from your ExchangeRateService, we normally have a Models or Contracts folder which is a good place for your CnbRatesResponse and Rate. Also a good idea to have sperate classes for the Client and its interface and put them in a separate Clients folder. Have a look at how it's done in www.github.com/Mews/ImaginaryProject. Restructuring this will also help you to adopt some of the SOLID principles

**Comment by 4hebek** [comment]
We should definitely add some tests to this, you can use XUnit to do that. A few happy path test cases with the required currencies i.e. USD, EUR, JPY, SAR, KES individually as well as testing getting them all in one call will be the minimum. You can then start thinking about the edge cases and unhappy paths for example for currency codes that aren't supported. These will act as foundations and help you programm your solution and hopefully help you cover all the exceptions mentioned above.

**Comment by 4hebek** [comment]
Before you put your work up for a review make sure it runs fine locally and you can call it successfully using Postman.

**Comment by 4hebek** [comment]
Please don't forget to update the readme, followed by all the documentation (e.g. on Confluence) as well as the Postman collection once this has been merged

**Comment by 4hebek** [comment]
A few things to address but a great job overall! Good use of minimal APIs, nice neat and readable solution, well done 😃 Let me know if you want to discuss any of the points in more detail, it's always easier to jump on a huddle on Slack.
